File: procedural/Queens.java
public class Queens {
    static int size = 0;
    static boolean[] row;
    static boolean[] up;
    static boolean[] down;
    static int[] board;

    public static void main(String[] args) {
        try {
            if (args.length > 0) size = Integer.valueOf(args[0]);
            if (size <= 0) return;
            row = new boolean[size];
            up = new boolean[2*size-1];
            down = new boolean[2*size-1];
            board = new int[size];
            queen(0);
        } catch (Exception e) {
        }
    }

    private static void queen(int n) {
        int i;

        if (n >= size) {
            for (i = 0; i < size-1; ++i) {
                System.out.printf("%d ", board[i]);
            }
            System.out.printf("%d\n", board[i]);
        } else {
            for (i = 0; i < size; ++i) {
                if (!row[i] && !up[n+i] && !down[n-i+size-1]) {
                    row[i] = up[n+i] = down[n-i+size-1] = true;
                    board[n] = i+1;
                    queen(n+1);
                    row[i] = up[n+i] = down[n-i+size-1] = false;
                }
            }
        }
    }
}

File: procedural/queens.scala
object Queens {
    def main(args: Array[String]) {
        val size = args(0).toInt
        val row = Array.fill(size)(true)
        val up = Array.fill(2 * size - 1)(true)
        val down = Array.fill(2 * size - 1)(true)
        val board = new Array[Int](size)
        var queen: (Int) => Unit = null
        queen = (n) => {
            if (n >= size) {
                for (i <- 0 until board.size - 1) print(board(i) + " ")
                println(board(board.size - 1))
            } else {
                for (i <- 0 until board.size) {
                    if (row(i) && up(n+i) && down(n-i+size-1)) {
                        row(i)=false; up(n+i)=false; down(n-i+size-1)=false
                        board(n) = i + 1
                        queen(n + 1)
                        row(i)=true; up(n+i)=true; down(n-i+size-1)=true
                    }
                }
            }
        }
        queen(0)
    }
}
